include/ck_tile/ops/epilogue/cshuffle_epilogue.hpp

```diff
@@ -1,194 +1,194 @@
+                          c_warp_y_lengths.get(number<NDimY - 1>{}));
+            return c_warp_y_lengths.get(number<NDimY - 1>{});
         }
-
-        // Apply the permutation to the tile coordinates
-        index_t permuted_tile_coords[CK_TILE_MAX_RANK];
-        for(index_t i = 0; i < kRank; ++i)
+        // M is contiguous dimension
+        else if constexpr(std::is_same_v<CLayout, tensor_layout::gemm::ColumnMajor>)
         {
-            permuted_tile_coords[i] = tile_coords[kPerm[i]];
-            // printf("The new permuted_tile_coords is: %d", permuted_tile_coords[i]);
+            // In this case each thread has just a single item in Mdim
+            return WG::WarpGemmAttribute::Impl::kCNLane / WG::kN;
         }
-
-        // Compute the permuted window origin
-        index_t permuted_window_origin[CK_TILE_MAX_RANK] = {0};
-        for(index_t i = 0; i < kRank; ++i)
+        else
         {
-            permuted_window_origin[i] = permuted_tile_coords[i] * tile_sizes[i];
-            // printf("The new permuted_window_origin is: %d", permuted_window_origin[i]);
+            static_assert(false, "Unsupported CLayout!");
         }
+    }
 
-        typename ODramWindowTmp::BottomTensorIndex step = {};
-        for(index_t i = 0; i < kRank; ++i)
-        {
-            step[i] = permuted_window_origin[i] - current_window_origin[i];
-        }
+    template <typename Problem>
+    CK_TILE_HOST_DEVICE static constexpr auto MakeLdsBlockDescriptor()
+    {
+        return make_naive_tensor_descriptor(
+            make_tuple(number<kMWave * kMPerXdl>{}, number<kNWave * kNPerXdl>{}),
+            make_tuple(number<kNWave * kNPerXdl>{}, number<1>{}));
+    }
 
-        // Move the window
-        move_tile_window(o_dram_window_tmp, step);
+    CK_TILE_HOST_DEVICE static constexpr index_t GetSmemSize()
+    {
+        return kMWave * kNWave * kMPerXdl * kNPerXdl * sizeof(ODataType);
+    }
 
-        // Permute the data within the tile if necessary
-        if constexpr(kTilePermute)
-        {
-            permute_tile_data(o_acc_tile);
-        }
+    template <typename ODramWindow,
+              typename OAccTile,
+              memory_operation_enum out_memory_data_op = memory_operation_enum::set>
+    CK_TILE_DEVICE auto
+    operator()(ODramWindow& out_dram_window, const OAccTile& o_acc_tile, void* p_smem)
+    {
 
-        // Store the tile data to the permuted location
-        if constexpr(kPadM || kPadN)
-        {
+        const index_t iMWarp = get_warp_id() / kNWave;
+        const index_t iNWarp = get_warp_id() - iMWarp * kNWave;
+
+        constexpr auto lds_block_desc = MakeLdsBlockDescriptor<Problem>();
+        auto o_lds_block              = make_tensor_view<address_space_enum::lds>(
+            static_cast<ODataType*>(p_smem), lds_block_desc);
+        auto in_lds_window =
+            make_tile_window(o_lds_block,
+                             make_tuple(number<kMPerXdl>{}, number<kNPerXdl>{}),
```

Comment:
Shouldn't you have in here: ?
```suggestion
                             make_tuple(number<kMWave * kMPerXdl>{}, number<kNWave * kNPerXdl>{}),
```

Reply:
No, this window is per Wavefront